tests/vmi_cloudinit_test.go

```diff
@@ -214,50 +195,42 @@ var _ = Describe("[rfe_id:151][crit:high][vendor:cnv-qe@redhat.com][level:compon
 				})
 			})
 
-			It("cloud-init instance-id should be stable", func() {
+			It("cloud-init instance-id should be stable", decorators.Conformance, func() {
```

Comment:
I would reconsider adding this one.
It is basically a test of cloudinit and the fact that KubeVirt has persistent storage. We start the VM twice just for that.

I'm not sure that it is needed at all.
It may make sense as an additional Expect for a VM started for another reason.
I'm certain it should not be in Conformance.

Reply:
>It is basically a test of cloudinit and the fact that KubeVirt has persistent storage. We start the VM twice just for that.

Not really, the instance-id is generated by platform (that is us).
So the test makes sure that on restart we do not generate different ID and therefore the cloudinit would not be re-triggered if you use persistent storage.

Reply:
I stand corrected. Where is the instance-id persisted? VM spec? I bet this is unit-testable.
I would not run a VM twice just for that, certainly not in `Conformance`.

Reply:
The implementation is kinda state-less and it can always change. That is why I want this, not having this covered makes cloud-init very much unusable.

Reply:
Would you educate me: where is the instance-id persisted?

The implementation cannot change if there is an explicit test that enforces it. If you unilaterally change the persistence, VMs that were booted in older versions would forget their identity unless you add an explicit upgrade mechanism.

Reply:
>Would you educate me: where is the instance-id persisted?

```
func cloudInitUUIDFromVMI(vmi *v1.VirtualMachineInstance) string {
	if vmi.Spec.Domain.Firmware == nil {
		return uuid.NewString()
	}
	return string(vmi.Spec.Domain.Firmware.UUID)
}
```

>The implementation cannot change if there is an explicit test that enforces it. If you unilaterally change the persistence, VMs that were booted in older versions would forget their identity unless you add an explicit upgrade mechanism.

That is correct but that is true for all other conformance tests.

Reply:
We seem to speak across each other... Let me retry.

`cloudInitUUIDFromVMI` generates a UUID for a VMI, but it does not persist it. To the contrary, it generates a random UUID for VMs without `vmi.Spec.Domain.Firmware`. I fail to find where we do persist the UUID.

I'm trying to say that having a persistent UUID is important regardless of cloud init and it merits a test of its own. I find testing it here by starting a VM twince as convoluted and too expensive. I suggest keeping this out of `Conformance` (for now?)

Reply:
Ah sorry, I should have include:
```
// setStableUUID makes sure the VirtualMachineInstance being started has a 'stable' UUID.
// The UUID is 'stable' if doesn't change across reboots.
func setupStableFirmwareUUID(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) {

	logger := log.Log.Object(vm)

	if vmi.Spec.Domain.Firmware == nil {
		vmi.Spec.Domain.Firmware = &virtv1.Firmware{}
	}

	existingUUID := vmi.Spec.Domain.Firmware.UUID
	if existingUUID != "" {
		logger.V(4).Infof("Using existing UUID '%s'", existingUUID)
		return
	}

	vmi.Spec.Domain.Firmware.UUID = types.UID(uuid.NewSHA1(firmwareUUIDns, []byte(vmi.ObjectMeta.Name)).String())
}
```
Without this it doesn't make sense... (me going to remove the useless argument)

> I find testing it here by starting a VM twince as convoluted and too expensive. I suggest keeping this out of Conformance (for now?)

@dankenigsberg 
How do you want to test this without restarting the VM? I am not sure I understand this.

Reply:
Thanks. So it seems that we do not persist the UUID, we just recompute it.

> How do you want to test this without restarting the VM? I am not sure I understand this

from the top of my head:
```
func GetVMUuid(vm) { ... }

vm = handCraftedVM()
Expect(GetStableFirmwareUuid(vm) = "7b11c052f9741792ef58f4c7ba8aa644c4fbfa0e")
```

Please note that unlike the current test, my suggested test would have stopped https://github.com/kubevirt/kubevirt/pull/12885 from being merged. I believe that current test is not good. It certainly does not merit being part of `Conformance`.

Reply:
I admit the test didn't catch #12885 but I am not sure your suggestion would do it as well. It seems the test would rely on the implementation details. 

A test that would start a VM in n-1 version, upgrade and restart the vm would be most explicit imho. 

>It certainly does not merit being part of Conformance.

Can you elaborate? Isn't this core attribute of cloud-init? Or do you point out that it is not good?

@acardace @fossedihelm Any arguments why you see this test as good enough not/for conformance?

Reply:
To me, it is good enough for conformance since it ensures that instance-id is not calculated every time the VM is started, and this is a hotpoint for cloudinit.
https://github.com/kubevirt/kubevirt/pull/12885 case is quite wider: the context there is that `instance-id` should be stable across upgrade.

Reply:
> it ensures that instance-id is not calculated every time the VM is started

Unfortunately, we do calculate Firmware.UUID every time the VM is started. We have a complicated test for ensuring we recalculate the same value every time the VM starts.

Keeping a stable Firmware.UUID is important indeed, but this is true regardless of cloud-init. Many other guest applications would be confused if the same VM boots with a different UUID. Why do we need cloudinit to verify that? Why do we need to run the VM twice? 

The Conformance test must be short quick and reliable. We should not add to it long and dubious tests. Historically, it is very hard to drop a test once it is added. Hence I'm asking to keep this test out of Conformance for now. Would you agree to introduce a follow-up PR explaining why this test is quick and important enough to be introduced to `Conformance`?

I think we'd be much better off if the VM controller would just add Firmware.UUID to vm.spec.template in case it is missing. That would be easy to unit test, and or included in an e2e test if you insist.

Reply:
>Keeping a stable Firmware.UUID is important indeed, but this is true regardless of cloud-init. Many other guest applications would be confused if the same VM boots with a different UUID. Why do we need cloudinit to verify that? Why do we need to run the VM twice?

The firmware UUID and cloud-init instance-id are two different things. We just use one for the other but I am not aware of any requirement to do it this way... Decoupling these in future without having separate test for each would cause breakages certainly...

There is commonality for sure, both needs to be persistent across reboots and across upgrades but I can't justify to have only 1 E2E test to cover both because the implementation is shared (as they are two different features). 

>The Conformance test must be short quick and reliable.

And is this test the opposite?

>I think we'd be much better off if the VM controller would just add Firmware.UUID to vm.spec.template in case it is missing. That would be easy to unit test, and or included in an e2e test if you insist.

Let's keep the implementation details out. Good E2E tests should be flexible to implementation changes, do we agree?

>Hence I'm asking to keep this test out of Conformance for now. Would you agree to introduce a follow-up PR explaining why this test is quick and important enough to be introduced to Conformance?

We certainly can but on what basis? I am happy to add the reasoning to the commit message if that is what we are missing, that is quit cheap for me.

Reply:
> 
> > The Conformance test must be short quick and reliable.
> 
> And is this test the opposite?

Indeed. It starts a VM twice (=slow and fragile). It depends heavily on a third-party software (cloud-init) instead of just testing the interface between us and them.

> 
> > Hence I'm asking to keep this test out of Conformance for now. Would you agree to introduce a follow-up PR explaining why this test is quick and important enough to be introduced to Conformance?
> 
> We certainly can but on what basis? I am happy to add the reasoning to the commit message if that is what we are missing, that is quit cheap for me.

Thanks. The basis is that this test is not good. IMHO it would be easy to contrive a simpler faster test that ensures that the blob we pass to cloud-init does not change. We must keep `Conformance` lean and short. Not even every good e2e test should be there, let alone dubious tests like this one.

Reply:
`The /var/lib/cloud/instance directory is a symbolic link that points to the most recenlty used instance-id directory. This folder contains the information cloud-init received from datasources, including vendor and user data. This can be helpful to review to ensure the correct data was passed.` See the last sentence, even the official doc suggest this is the right way.

Reply:
@dankenigsberg let me know how can we move forward here. Happy to bring this up with wider forum...

Reply:
I'd be happy if you not add `decorators.Conformance` to this test, and continue the discussion in a follow-up PR.

Reply:
@dankenigsberg PTAL

Reply:
Now that test_id:3186 is kept out of Conformance, I see this PR as a net positive.

Thank you for this cleanup.

/lgtm

Thanks, @EdDev, for letting this thread conclude. I believe there is no pending discussion, hence

/hold cancel